pkg/snapstore/s3_snapstore.go

```diff
@@ -512,35 +517,115 @@ func (s *S3SnapStore) partUploader(wg *sync.WaitGroup, stopCh <-chan struct{}, s
 	}
 }
 
-// List will return sorted list with all snapshot files on store.
+// List returns a sorted list of snapshot files present in the object store.
+// If Bucket versioning is not enabled for S3 bucket:
+//   - It returns a sorted list of all snapshot files present in the object store.
+//
+// If Bucket versioning is enabled for S3 bucket:
+//   - It returns a sorted list of all latest snapshot files present in the object store.
+//   - It also captures the "ImmutabilityExpiryTime" and "VersionID" of corresponding versioned snapshot.
 func (s *S3SnapStore) List(_ bool) (brtypes.SnapList, error) {
+	var snapList brtypes.SnapList
 	prefixTokens := strings.Split(s.prefix, "/")
 	// Last element of the tokens is backup version
 	// Consider the parent of the backup version level (Required for Backward Compatibility)
 	prefix := path.Join(strings.Join(prefixTokens[:len(prefixTokens)-1], "/"))
 
-	var snapList brtypes.SnapList
-	in := &s3.ListObjectsInput{
-		Bucket: aws.String(s.bucket),
-		Prefix: aws.String(prefix),
-	}
-	err := s.client.ListObjectsPages(in, func(page *s3.ListObjectsOutput, lastPage bool) bool {
-		for _, key := range page.Contents {
-			k := (*key.Key)[len(*page.Prefix):]
-			if strings.Contains(k, backupVersionV1) || strings.Contains(k, backupVersionV2) {
-				snap, err := ParseSnapshot(path.Join(prefix, k))
-				if err != nil {
-					// Warning
-					logrus.Warnf("Invalid snapshot found. Ignoring it: %s", k)
+	// Get the status of bucket versioning.
+	// Note: Bucket versioning will always be enabled for object lock.
+	versioningStatus, err := s.client.GetBucketVersioning(&s3.GetBucketVersioningInput{
+		Bucket: &s.bucket,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if versioningStatus.Status != nil && *versioningStatus.Status == s3.BucketVersioningStatusEnabled {
+		// versioning is found to be enabled on given bucket.
+		logrus.Info("Bucket versioning is found to be enabled.")
+
+		var bucketImmutableExpiryTimeInDays *int64
+		var isObjectLockEnabled bool
+
+		if err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
+			return err != nil
+		}, func() error {
+			if isObjectLockEnabled, bucketImmutableExpiryTimeInDays, err = GetBucketImmutabilityTime(s); err != nil {
+				logrus.Errorf("unable to check bucket immutability retention period: %v", err)
+				return err
+			}
+			if !isObjectLockEnabled {
+				logrus.Warnf("Bucket versioning is found to be enabled but object lock is not found to be enabled.")
+				logrus.Warnf("Please enable the object lock for the given bucket to to have immutable snapshots.")
+			}
+			if isObjectLockEnabled && bucketImmutableExpiryTimeInDays == nil {
+				logrus.Warnf("Object lock is found to be enabled but object lock rules or default retention period are not found to be set.")
+			}
+			return nil
```

Comment:
```suggestion
			if isObjectLockEnabled, bucketImmutableExpiryTimeInDays, err = GetBucketImmutabilityTime(s); err != nil {
				logrus.Errorf("unable to check bucket immutability retention period: %v", err)
				return err
			} else if !isObjectLockEnabled {
				logrus.Warnf("Bucket versioning is found to be enabled but object lock is not found to be enabled.")
				logrus.Warnf("Please enable the object lock for the given bucket to to have immutable snapshots.")
			} else if bucketImmutableExpiryTimeInDays == nil {
				logrus.Warnf("Object lock is found to be enabled but object lock rules or default retention period are not found to be set.")
			}
			return nil
```

Just to make sure the condition `isObjectLockEnabled` is checked only once. Feel free to ignore if you feel it doesn't matter.

Reply:
I have made the changes, PTAL